### PR TITLE
Fixed #17002 -- M2M traversal through models using concrete inheritance

### PR DESCRIPTION
**Pull request by InvalidInterrupt**
https://code.djangoproject.com/ticket/17002

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, add to whitelist.

File: django/db/models/options.py [hunk 1]
```diff
@@ -667,6 +668,28 @@ def get_ancestor_link(self, ancestor):
                 # links
                 return self.parents[parent] or parent_link
 
+    def get_path_to_parent(self, parent):
+        """
+        Get path from current model to a parent model.
```
Comment (timgraham) [inline django/db/models/options.py]: Suggested wording: "Return a list of PathInfos containing the path from the current model to a parent model, or an empty list if...."

Could you also describe the circumstances where the list will be length > 1?

File: django/db/models/fields/related.py [hunk 2]
```diff
@@ -1525,10 +1525,27 @@ def _get_path_info(self, direct=False):
         if direct:
             join1infos = linkfield1.get_reverse_path_info()
             join2infos = linkfield2.get_path_info()
+            join1_final = join1infos[-1].to_opts
```
Comment (timgraham) [inline django/db/models/fields/related.py]: Maybe some inline comments for this code?

File: django/db/models/options.py [hunk 3]
```diff
@@ -690,6 +690,24 @@ def get_path_to_parent(self, parent):
...
+        Returns an empty list if parent is not a parent of the current model.
+        """
+        if self.model is parent:
+            return []
+        model = self.concrete_model
+        chain = model._meta.get_base_chain(parent)
+        chain.reverse()
+        chain.append(model)
+        path = []
+        for i in range(len(chain) - 1):
```
Comment (timgraham) [inline django/db/models/options.py]: Some inlines comments? 
Not sure if something like `for i, obj in enumerate(chain[:-1])` might be a better formulation of this logic.

File: tests/m2m_through_regress/models.py [hunk 4]
```diff
@@ -94,3 +94,31 @@ class CarDriver(models.Model):
 
     def __str__(self):
         return "pk=%s car=%s driver=%s" % (str(self.pk), self.car, self.driver)
+
+
+# Through models using multi-table inheritance
+class Event(models.Model):
+    name = models.CharField(max_length=50, unique=True)
+    people = models.ManyToManyField('Person', through='IndividualCompetitor')
+    special_people = models.ManyToManyField('Person', through='ProxiedIndividualCompetitor',
```
Comment (timgraham) [inline tests/m2m_through_regress/models.py]: Use hanging indent:
```
special_people = models.ManyToManyField(
    'Person',
    through='ProxiedIndividualCompetitor',
    related_name='special_event_set',
)
```

File: tests/m2m_through_regress/tests.py [hunk 5]
```diff
@@ -268,3 +268,37 @@ def test_sequence_creation(self):
             ': 1}}, {"pk": 1, "model": "m2m_through_regress.person", "fields": {"name": "Guido"}}, {"pk": 1, '
             '"model": "m2m_through_regress.group", "fields": {"name": "Python Core Group"}}]'
         )
+
+
+class M2MThroughMultiTableTestCase(TestCase):
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: A separate test file, e.g. `test_multitableinheritance.py` might be nice.

File: tests/m2m_through_regress/models.py [hunk 6]
```diff
@@ -94,3 +94,31 @@ class CarDriver(models.Model):
...
+                                            related_name='special_event_set')
+    teams = models.ManyToManyField('Group', through='CompetingTeam')
+
+
+class Competitor(models.Model):
+    event = models.ForeignKey('Event', models.CASCADE)
+    rank = models.IntegerField(blank=True, null=True)
+    disqualified = models.BooleanField()
```
Comment (timgraham) [inline tests/m2m_through_regress/models.py]: It's not clear to me if some of the extra fields like `disqualified` are needed. In general, it's nice to keep things minimal.

File: tests/m2m_through_regress/tests.py [hunk 7]
```diff
@@ -302,3 +302,13 @@ def test_m2m_reverse_query_proxied(self):
         result = self.chris.special_event_set.all()
         self.assertEqual(len(result), 1)
         self.assertEqual(result[0], self.event)
+
+    def test_m2m_prefetch_proxied(self):
+        result = Event.objects.filter(name='Exposition Match').prefetch_related('special_people')
+        # Trigger Queryset evaluation
+        str(result)
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: Would probably be nice to check the results and the proper number of queries.

File: django/db/models/options.py (hunk 1 above)
Comment (InvalidInterrupt) [inline django/db/models/options.py]: I'm not very familiar with the inner workings of the Django ORM, but it seems to me that the list would contain multiple entries if the chosen parent is a grandparent or more distant model, using multi-table inheritance.

**Comment by timgraham** [comment]
To merge this as multiple commits (if you think that adds value), it was be nice to have the review fixes incorporated into the first three commits rather than as a separate commit. If you envisioned the commits to be squashed when merging, let me know. I don't have strong feeling either way.

**Comment by InvalidInterrupt** [comment]
I was developing under the assumption it would be squashed when merging. If you don't feel strongly it should be multiple commits, then I think it should be squashed.

**Comment by timgraham** [comment]
merged in 98359109eb0ed68a5821476bcd797455723aaaba, thanks!

**Comment by timgraham** [comment]
I noticed some check errors in the models:

m2m_through_regress.Event.people: (fields.E340) The field's intermediary table 'm2m_through_regress_individualcompetitor' clashes with the table name of 'm2m_through_regress.ProxiedIndividualCompetitor'.
m2m_through_regress.Event.special_people: (fields.E340) The field's intermediary table 'm2m_through_regress_individualcompetitor' clashes with the table name of 'm2m_through_regress.IndividualCompetitor'.

How should we resolve them?

**Comment by InvalidInterrupt** [comment]
Having a M2M which uses a proxy of an existing through models doesn't seem all that useful to me, but assuming we want it to work if anyone tries it looks like https://github.com/django/django/blob/98359109eb0ed68a5821476bcd797455723aaaba/django/db/models/fields/related.py#L1450 could be changed to `if m2m_db_table in registered_tables and registered_tables[m2m_db_Table]._meta.concrete_model != self.remote_field.through._meta.concrete_model:`.
Unfortunately I don't have enough time to spare at the moment to try it myself.

**Comment by timgraham** [comment]
Thanks, I created https://github.com/django/django/pull/7753 with that suggestion.

**Pull request closed without merging**